### PR TITLE
Feature robo

### PR DESCRIPTION
**Pull request by daniel-samson**
## Description

Adding the [Robo](https://robo.li/) tasks to SuiteCRM.

## Motivation and Context

Robo tasks provides a consistent command line interface to automate common tasks in PHP.

## How To Test This

- run composer update
- run ./vendor/bin/robo

To set up environment variables
- run ./vendor/bin/robo  configure:tests

To build Sass (SuiteP):
- run ./vendor/bin/robo  build:suitep

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

File: RoboFile.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ */
+require_once __DIR__.'/vendor/autoload.php';
+/**
+ * This is project's console commands configuration for Robo task runner.
+ *
+ * @see http://robo.li/
+ */
+class RoboFile extends \Robo\Tasks
+{
+    use \SuiteCRM\Robo\OperatingSystemTrait;
```
Comment (gymad) [inline RoboFile.php]: Do not use trait even if it's unnecessary. You are using these traits only one place, try to replace it to classes as dependency or helper. 

> Benefits of traits is that you reduce code duplication whilst preventing complicated class inheritance that might not make sense within the context of your application.

> Traits are not the answer to all of your problems. Using a Trait in the wrong situation is most definitely a bad decision. If you are trying to crowbar the functionality you desire into a class using Traits, you are probably doing something wrong.

a very good explanation of traits I found with a quick google search:
http://www.culttt.com/2014/06/25/php-traits/

Comment (zoltankocsardi) [reply RoboFile.php]: I agree, it's not necessary to use custom traits here.

BUT

I'd suggest we should separate tasks if they are more complex. According to the [robo specs](https://robo.li/extending/):

> Note: The Tasks traits are called loadTasks in Robo core.

So we can extend this with a new loadTask trait and can use this here.

File: RoboFile.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    // define public methods as commands
+
+    /**
+     * configure environment for testing
+     * @see https://docs.suitecrm.com/developer/appendix-c---automated-testing/#_environment_variables
+     * @param array $opts optional command line arguments
+     */
+    public function configureTests(
+        array $opts = [
+            'database_driver' => '',
```
Comment (zoltankocsardi) [inline RoboFile.php]: Please, get these attributes from the config.php.
[implement a new method to get the config].

File: RoboFile.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+        // SuiteCRM Instance
+        $this->askDefaultOptionWhenEmpty('Instance URL:', 'http://localhost', $opts['instance_url']);
+        $this->askDefaultOptionWhenEmpty('Instance Admin Username:', 'admin', $opts['instance_admin_user']);
+        $this->askDefaultOptionWhenEmpty('Instance Admin Password:', 'admin1', $opts['instance_admin_password']);
+        $this->askDefaultOptionWhenEmpty('Instance OAuth2 Client ID:', 'suitecrm_client', $opts['instance_client_id']);
+        $this->askDefaultOptionWhenEmpty('Instance OAuth2 Client Secret:', 'secret', $opts['instance_client_secret']);
+
+        if ($this->isOsWindows()) {
+            $this->say('Windows detected');
```
Comment (zoltankocsardi) [inline RoboFile.php]: Just a hint. You can use **say** after a task ran successfully.
Example:

`
if ($this->taskExec("rm -rf whateva)->run()->wasSuccessful()) {
    $this->say('Removed whateva');
}
`

File: lib/Robo/SassTrait.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+trait SassTrait
+{
+    use \SuiteCRM\Robo\PathTrait;
+    /**
+     * @param string $colorScheme eg Dawn
+     */
+    private function buildSuitePColorScheme($colorScheme)
+    {
+        $this->_exec(
+            $this->toOsPath(
```
Comment (zoltankocsardi) [inline lib/Robo/SassTrait.php]: Like I mentioned before, you can replace this trait easily with a simple command
$this->taskExec($command)->run();

File: RoboFile.php (hunk 3 above)
Comment (gymad) [reply RoboFile.php]: "whateva"

File: lib/Utility/OperatingSystem.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        return stristr(php_uname('s'), 'Windows') !== FALSE;
+    }
+
+    /**
+     * @param string $path
+     * @return string path converted for operating system eg Linux, Mac OS, Windows
+     */
+    public function toOsPath($path)
+    {
+        return str_replace('/', DIRECTORY_SEPARATOR, $path);
```
Comment (gymad) [inline lib/Utility/OperatingSystem.php]: I assuming you wanted a correction on path separators, converting for the current OS?
`$path = preg_replace('/[\\\\\/]/', DIRECTORY_SEPARATOR, $path);`

Comment (daniel-samson) [reply lib/Utility/OperatingSystem.php]: yes.this method converts the unix convention ./vendor/bin/codecept to .\vendor\bin\codecept on Windows other wise the path with remain the same.

Comment (gymad) [reply lib/Utility/OperatingSystem.php]: Ok. In theory: In my windows machine the vendor path is for e.g. `vendor\bin\codecept\`. What is the return value if I give it to your function in a linux machine?

File: lib/Robo/Plugin/Commands/TestEnvironmentCommands.php [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+{
+    use \SuiteCRM\Robo\Traits\RoboTrait;
+
+    /**
+     * Configure environment for testing
+     * @see https://docs.suitecrm.com/developer/appendix-c---automated-testing/#_environment_variables
+     * @param array $opts optional command line arguments
+     */
+    public function configureTests(
+        array $opts = [
```
Comment (gymad) [inline lib/Robo/Plugin/Commands/TestEnvironmentCommands.php]: array $opts = [..]? is it a correct syntax?

File: tests/unit/lib/SuiteCRM/Utility/OperatingSystemTest.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+
+class OperatingSystemTest extends \Codeception\Test\Unit
```
[the remark this answers is not in the document]
Comment (daniel-samson) [reply tests/unit/lib/SuiteCRM/Utility/OperatingSystemTest.php]: I need the dependency injection for this class, so I can't extend but I can use the same traits. It seems to work.

[the remark this answers is not in the document]
Comment (gymad) [reply tests/unit/lib/SuiteCRM/Utility/OperatingSystemTest.php]: just replace `\Codeception\Test\Unit` => `SuiteCRM\StateCheckerUnitAbstract`

File: tests/unit/lib/SuiteCRM/Utility/OperatingSystemTest.php [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+{
+    use \SuiteCRM\StateCheckerTrait;
```
Comment (gymad) [inline tests/unit/lib/SuiteCRM/Utility/OperatingSystemTest.php]: can you just replace \Codeception\Test\Unit => SuiteCRM\StateCheckerUnitAbstract

**Comment by Dillon-Brown** [comment]
Assessed

**Comment by Dillon-Brown** [comment]
Assessed

File: tests/unit/lib/SuiteCRM/Utility/OperatingSystemTest.php [hunk 9]
```diff
@@ -1,10 +1,8 @@
 <?php
 
 
-class OperatingSystemTest extends \Codeception\Test\Unit
+class OperatingSystemTest extends \SuiteCRM\StateCheckerUnitAbstract
```
Comment (gymad) [inline tests/unit/lib/SuiteCRM/Utility/OperatingSystemTest.php]: Thanks

**Comment by connorshea** [comment]
Is there documentation on adding a custom Robo task to SuiteCRM? I've got a task at `custom/lib/Robo/Plugin/Commands/CustomCommands.php` that looks like this:

```php
<?php
namespace SuiteCRM\Custom\Robo\Plugin\Commands;
use Robo\Task\Base\loadTasks;

class CustomCommands extends \Robo\Tasks
{
    use loadTasks;

    /**
     * This Robo task will deploy a given branch.
     *
     * @param array $opts
     * @option $branch The branch to deploy, e.g. master, development, etc.
     * @usage deploy --branch=master
     */
    public function deploy(array $opts = ['branch' => 'master'])
    {
        $this->say("Hello, world");
        return Robo\Result::success($this, "Deployed successfully.");
    }
}
```

But `.vendor/bin/robo list` doesn't seem to include my deploy task. Anything I can do to make this work?

**Comment by gymad** [comment]
( @connorshea FYI: `composer dump-autoload` does the job faster. )

**Pull request merged**